lightning/src/onion_message/messenger.rs

```diff
@@ -598,19 +633,27 @@ where
 		features.set_onion_messages_optional();
 		features
 	}
-}
 
-impl<ES: Deref, NS: Deref, L: Deref, MR: Deref, OMH: Deref, CMH: Deref> OnionMessageProvider
-for OnionMessenger<ES, NS, L, MR, OMH, CMH>
-where
-	ES::Target: EntropySource,
-	NS::Target: NodeSigner,
-	L::Target: Logger,
-	MR::Target: MessageRouter,
-	OMH::Target: OffersMessageHandler,
-	CMH::Target: CustomOnionMessageHandler,
-{
-	fn next_onion_message_for_peer(&self, peer_node_id: PublicKey) -> Option<msgs::OnionMessage> {
+	// Before returning any messages to send for the peer, this method will see if any messages were
+	// enqueued in the handler by users, find a path to the corresponding blinded path's introduction
+	// node, and then enqueue the message for sending to the first peer in the full path.
+	fn next_onion_message_for_peer(&self, peer_node_id: PublicKey) -> Option<OnionMessage> {
+		// Enqueue any initiating `OffersMessage`s to send.
+		for message in self.offers_handler.release_pending_messages() {
+			let PendingOnionMessage { contents, destination, reply_path } = message;
+			self.find_path_and_enqueue_onion_message(
+				contents, destination, reply_path, format_args!("when sending OffersMessage")
+			);
+		}
+
+		// Enqueue any initiating `CustomMessage`s to send.
+		for message in self.custom_handler.release_pending_custom_messages() {
+			let PendingOnionMessage { contents, destination, reply_path } = message;
+			self.find_path_and_enqueue_onion_message(
+				contents, destination, reply_path, format_args!("when sending CustomMessage")
+			);
+		}
```

Comment:
I chatted with @TheBlueMatt about this in person, awhile back. IIRC, this is essentially a user-initiated action, so it would be rare. The alternative would be needing to initiate path finding in handler implementations, which would require them having a router.

But you're right that it could lead to a delay in message processing. @TheBlueMatt Is ten messages too many? Should we re-consider calling path finding here?

Reply:
Ah okay, yeah I suppose it *is* a user-initiated action so it should be rare. Just struck me as weird when reading it.